### ISSUE TITLE
Custom Runtime build on NodeJS

### ISSUE THREAD

**Issue by thanhtung6824** [paritytech/substrate#7995]
I've read the docs and see Substrate allow to create custom runtime by any language target to Wasm. I want to build my own custom runtime base on NodeJs.
Is there a document for that? Need some helps

**Comment by bkchr** [comment]
CC @Stefie

**Comment by thanhtung6824** [comment]
I try to implement like: Using consensus, p2p FRAME handles all the p2p and consensus stuff then write custom Pallet. It's will forward transaction when need to be validated, or when they're ready to be committed to a block. It's seems like Tendermint Core and Abci application. Is this approach possible?

**Comment by Stefie** [comment]
@thanhtung6824  Can you point me to which part of the docs you're referring to so I can get a better understanding of what you're trying to achieve?

In general you can't write a Substrate runtime in NodeJS because you can't compile Node to WebAssembly. The only JavaScript-flavoured language that can be compiled to Wasm is [AssemblyScript](https://www.assemblyscript.org/).
The people from Limechain are working on a custom runtime implementation in AssemblyScript which you can find in the [Subsembly repository](https://github.com/LimeChain/subsembly). You should have a look at it!

> I try to implement like: Using FRAME handles all the p2p and consensus stuff then write custom Pallet

Is your idea to mix the Substrate runtime generated by the Rust code with a custom runtime code written in a different language (NodeJS in your case)? I'm afraid that that is not possible. If you want to write your own runtime in a different language, you need to rewrite the whole thing. Mixing two languages is not possible because you need to end up with only one Wasm binary.

**Comment by thanhtung6824** [comment]
@Stefie 
Thank you for reply Stefie, My Idea is write the Substrate runtime in Rust code, it's handles all the p2p and consensus stuff and just forwards transactions to the my runtime (call api to my runtime in NodeJS) when need to be validated, or when they're ready to be committed to a block. 
Now my chain is running with Tendermint Core and I want migrate to Substrate with less code change as much as possible

**Comment by Stefie** [comment]
Ah, ok, I get it! I'm afraid I'm not the right person to answer that. Maybe @danforbes can help?

**Comment by thanhtung6824** [comment]
@danforbes Can you help?
Can't wait to build some exciting app on Substrate

**Comment by shawntabrizi** [comment]
Technically speaking you should be able to create a Substrate runtime using any programming language which can compile to Wasm, however, practically speaking it would be very hard to do this with any language other than Rust for the moment.

Basically the only requirement of the runtime is to satisfy the Runtime API that the Substrate Core Client expects.

However, there is no such template or setup for any of that in NodeJS or any JS today. So if you wanted to do that, you must build all those things from scratch, and likely you would be diverging far from anything that resembles Tendermint Core.

**Comment by kianenigma** [comment]
> However, there is no such template or setup for any of that in NodeJS or any JS today. So if you wanted to do that, you must build all those things from scratch, and likely you would be diverging far from anything that resembles Tendermint Core.

Indeed. I think the closest thing that we have to what you want to do is what @JoshOrndorff worked on (but I belive was not published anywhere). https://github.com/substrate-developer-hub/recipes/tree/joshy-frameless-runtime

This is how a runtime looks like without Frame, in pure Rust. Porting this to any other language will be slightly easier, but you'd still need to figure out the details of the rutnime/wasm APIs in your target language. 

Also, note that this runtime is doing absolutely nothing. Doing anything useful would, again, make it substantially more complicated.

**Comment by JoshOrndorff** [comment]
```diff
- This is how a runtime looks like without Frame, in pure Rust.
+ This is how a not-fully-working runtime looks like without Frame, in pure Rust.
```

I only ever got it to author empty blocks. It couldn't handle transactions properly. This stuff is hard.

**Comment by JoshOrndorff** [comment]
I'll just report back here in case anyone discovers this issue in the future.

The runtime without FRAME is now fully working. We have example code along with a few exercises for people who want to get familiar with it. https://github.com/JoshOrndorff/frameless-node-template

There is also now Tuxedo, which is an alternative to FRAME based on the UTXO model. It isn't quite production ready yet, but if you want to see an example of how to work without FRAME, it is great for that. https://github.com/Off-Narrative-Labs/Tuxedo/

**Comment by bkchr** [comment]
Ty for the update @JoshOrndorff!
